addons/stock_account/wizard/stock_valuation_layer_revaluation.py

```diff
@@ -0,0 +1,146 @@
+            else:
+                reval.new_value_by_qty = 0.0
+
+    def action_validate_revaluation(self):
+        """ Revaluate the stock for `self.product_id` in `self.company_id`.
+
+        - Change the stardard price with the new valuation by product unit.
+        - Create a manual stock valuation layer with the `added_value` of `self`.
+        - If the Inventory Valuation of the product category is automated, create
+        related account move.
+        """
+        self.ensure_one()
+        if self.currency_id.is_zero(self.added_value):
+            raise UserError(_("The added value doesn't have any impact on the stock valuation"))
+
+        product_property_cost_method = self.product_id.with_company(self.company_id).categ_id.property_cost_method
+
+        remaining_svls = self.env['stock.valuation.layer'].search([
+            ('product_id', '=', self.product_id.id),
+            ('remaining_qty', '>', 0),
+            ('company_id', '=', self.company_id.id),
+        ])
+
+        # Create a manual stock valuation layer
+        svl_vals = {
+            'company_id': self.company_id.id,
+            'product_id': self.product_id.id,
+            'description': _("Manual Stock Valuation: %s.%s") % (
+                self.reason or _("No Reason Given"),
+                product_property_cost_method == 'average' and (_(" Product cost updated from %s to %s.") % (self.product_id.standard_price, self.product_id.standard_price + self.added_value / self.current_quantity_svl)) or ''),
+            'value': self.added_value,
+            'quantity': 0,
+        }
+
+        remaining_qty = sum(remaining_svls.mapped('remaining_qty'))
+        remaining_value = self.added_value
+        remaining_value_unit_cost = self.currency_id.round(remaining_value / remaining_qty)
+        for svl in remaining_svls:
+            if float_is_zero(svl.remaining_qty - remaining_qty, precision_rounding=self.product_id.uom_id.rounding):
+                svl.remaining_value += remaining_value
+            else:
+                taken_remaining_value = remaining_value_unit_cost * svl.remaining_qty
+                svl.remaining_value += taken_remaining_value
+                remaining_value -= taken_remaining_value
+                remaining_qty -= svl.remaining_qty
+
+        stock_valuation_layer = self.env['stock.valuation.layer'].sudo().create(svl_vals)
+
+        # Update the stardard price in case of AVCO
+        if product_property_cost_method == 'average':
+            self.product_id.with_context(disable_auto_svl=True).standard_price += self.added_value / self.current_quantity_svl
+
+        if self.property_valuation != 'real_time':
+            return True
+
+        # If the Inventory Valuation of the product category is automated, create related account move.
+        if not self.env['stock.valuation.layer'].check_access_rights('read', raise_exception=False):
```

Comment:
account move

Reply:
Well, I prefer to mimic the behavior of the landed cost for coherence. it means, no check and no sudo, let manage the security of account move by itself. What do you think ?

Reply:
OK